### PR TITLE
[core] Cancel pending dependency resolution before failing a task

### LINKED ISSUE #18908: [core] Assertion failure (it->second.submitted_task_ref_count > 0) during dependency resolution
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Core

### What happened + What you expected to happen

```
[2021-09-25 13:34:55,450 C 655 718] reference_count.cc:352:  Check failed: it->second.submitted_task_ref_count > 0
--
  | *** StackTrace Information ***
  | ray::SpdLogMessage::Flush()
  | ray::RayLog::~RayLog()
  | ray::core::ReferenceCounter::RemoveSubmittedTaskReferences()
  | ray::core::ReferenceCounter::UpdateSubmittedTaskReferences()
  | ray::core::TaskManager::OnTaskDependenciesInlined()
  | std::_Function_handler<>::_M_invoke()
  | ray::core::CoreWorkerMemoryStore::Put()
  | ray::core::TaskManager::MarkPendingTaskFailed()
  | ray::core::TaskManager::PendingTaskFailed()
  | ray::core::CoreWorkerDirectActorTaskSubmitter::DisconnectActor()
  | ray::core::ActorManager::HandleActorStateNotification()
  | std::_Function_handler<>::_M_invoke()
  | ray::gcs::GcsPubSub::ExecuteCommandIfPossible()::{lambda()#1}::operator()()
  | std::_Function_handler<>::_M_invoke()
  | std::_Function_handler<>::_M_invoke()
  | boost::asio::detail::completion_handler<>::do_complete()
  | boost::asio::detail::scheduler::do_run_one()
  | boost::asio::detail::scheduler::run()
  | boost::asio::io_context::run()
  | ray::core::CoreWorker::RunIOService()
  | execute_native_thread_routine
```

### Reproduction script

https://buildkite.com/ray-project/periodic-ci/builds/1112#e3ad94ef-d92f-4b57-9759-e3693cda9884

### Anything else

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

Actor tasks are sometimes failed while their dependencies are still being resolved. This can cause hanging or crashes when we resolve the dependencies for a task that has already been canceled. It can lead to a crash from the ref counter when, for the same actor, actor task 1 depends on actor task 2. The sequence is:
1. Actor tasks 1 and 2 queued, 1 depends on 2.
2. Fail actor task 1. We clear its refs, including its dependency on 2.
3. Fail actor task 2. We store an error as its return value. Since task 1 depends on it, we inline the dependency and try to clear task 1's refs again, causing a ref counting error because we already cleared them in step 2.

This PR fixes the issue by canceling dependency resolution for tasks before failing them. This involves some refactoring of the LocalDependencyResolver. Most of the changes are for testing (split out the unit tests for LocalDependencyResolver into their own suite).

## Related issue number

Closes #18908.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/core_worker/transport/dependency_resolver.cc [hunk 1]
```diff
@@ -105,52 +86,80 @@ void LocalDependencyResolver::ResolveDependencies(
...
         });
   }
 
   for (const auto &actor_id : state->actor_dependencies) {
     actor_creator_.AsyncWaitForActorRegisterFinish(
-        actor_id, [this, state, on_complete](const Status &status) {
-          if (!status.ok()) {
-            state->status = status;
+        actor_id, [this, task_id, on_complete](const Status &status) {
+          std::unique_ptr<TaskState> resolved_task_state = nullptr;
```
Comment (scv119) [inline src/ray/core_worker/transport/dependency_resolver.cc:141]: there are some duplications for both `AsyncWaitForActorRegisterFinish` and `GetAsync`. Is is possible to unify them?

**Comment by scv119** [comment]
looks very good! mostly just nits

**Comment by scv119** [comment]
tests failure looks related: //:dependency_resolver_test
```


(18:50:25) [5,172 / 5,182] 80 / 86 tests, 1 failed; Compiling src/ray/raylet/placement_group_resource_manager_test.cc; 46s local, remote-cache ... (10 actions, 9 running)
--
  | (18:50:26) FAIL: //:dependency_resolver_test (see /root/.cache/bazel/_bazel_root/5fe90af4e7d1ed9fcf52f59e39e126f5/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/dependency_resolver_test/test_attempts/attempt_1.log)
  | (18:50:26) FAIL: //:dependency_resolver_test (see /root/.cache/bazel/_bazel_root/5fe90af4e7d1ed9fcf52f59e39e126f5/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/dependency_resolver_test/test_attempts/attempt_2.log)
  | (18:50:26) FAIL: //:dependency_resolver_test (see /root/.cache/bazel/_bazel_root/5fe90af4e7d1ed9fcf52f59e39e126f5/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/dependency_resolver_test/test.log)


```

File: src/ray/core_worker/transport/dependency_resolver.cc (hunk 1 above)
Comment (stephanie-wang) [reply src/ray/core_worker/transport/dependency_resolver.cc:142]: Hmm I couldn't figure out a way to do this nicely.

**Review by scv119 — APPROVED** [review]

File: src/ray/core_worker/transport/direct_actor_task_submitter.cc [hunk 2]
```diff
@@ -484,12 +487,15 @@ void CoreWorkerDirectActorTaskSubmitter::HandlePushTaskReply(
     RAY_CHECK(queue_pair != client_queues_.end());
     auto &queue = queue_pair->second;
 
+    // If the actor is already dead, immediately mark the task object is failed.
```
Comment (jjyao) [inline src/ray/core_worker/transport/direct_actor_task_submitter.cc:490]: nit: as failed

**Review by jjyao — APPROVED** [review]

**Pull request merged**